### PR TITLE
Add DeepSeek-V3 MXFP8 + DeepEP pre-training recipe

### PR DESCRIPTION
**Pull request by HoomanRamezani**
Add DeepSeek-V3 pre-training recipes (671B and 16B MoE) for B200 clusters with TorchTitan. 
Includes Slurm scripts and config to reproduce the results from accompanying PyTorch blog post.

### CODE REVIEW

File: deepep/README.md [hunk 1]
```diff
@@ -61,12 +61,19 @@ pip install nvidia-nvshmem-cu13
...
-git clone https://github.com/deepseek-ai/DeepEP.git
-cd DeepEP
+# Build dependencies
+sudo apt install -y python3-dev   # Python.h headers for C extension compilation
+pip install ninja
+
+# NVSHMEM setup
+export NVSHMEM_DIR=$(python3 -c "import nvidia.nvshmem; print(nvidia.nvshmem.__path__[0])")
+export TORCH_CUDA_ARCH_LIST="10.0"  # Use 9.0 for Hopper (H100/H200)
+ln -s libnvshmem_host.so.3 "$NVSHMEM_DIR/lib/libnvshmem_host.so"  # Symlink
```
Comment (Maurits-de-Groot) [inline deepep/README.md]: consider either changing to ls -sf to overwrite (with `-f` flag) in case the user has make a mistake or checking if it already exists with `[ -e "$NVSHMEM_DIR/lib/libnvshmem_host.so" ]` and if not (`||`) writing it to prevent overwriting (so complete line would be `[ -e "$NVSHMEM_DIR/lib/libnvshmem_host.so" ] || ln -s libnvshmem_host.so.3 "$NVSHMEM_DIR/lib/libnvshmem_host.so"`

File: deepep/README.md [hunk 2]
```diff
@@ -61,12 +61,19 @@ pip install nvidia-nvshmem-cu13
...
 
-NVSHMEM_DIR='<path to nvshmem install>' \
-TORCH_CUDA_ARCH_LIST="10.0" \  # Use 9.0 for Hopper (H100/H200)
-python setup.py build && python setup.py install
+# Build and install
+git clone https://github.com/deepseek-ai/DeepEP.git
+cd DeepEP && git checkout 29d31c0
+python3 setup.py build && python3 setup.py install
```
Comment (Maurits-de-Groot) [inline deepep/README.md:76]: in https://github.com/deepseek-ai/DeepEP setup build is only used for development - do we need this/have we tested without?

File: pytorch-dsv3-mxfp8/README.md [hunk 3]
```diff
@@ -0,0 +1,73 @@
...
+
+The recipes include MXFP8 mixed-precision training via TorchAO and DeepEP-accelerated expert communication, which together deliver up to 41% faster training throughput over BF16 without degrading convergence. For the full results and technical deep-dive, see the accompanying PyTorch blog post **[[Link Coming Soon]()]**.
+
+## Prerequisites
+
+- Access to a Nebius Soperator cluster:
+  - 32 nodes, 8x NVIDIA B200 GPUs per node, NVLink/NVSwitch and InfiniBand interconnects
+  - Shared filesystem mounted at `/mnt/data`
+- Python 3.12+
+- CUDA 12.8 toolkit on compute nodes
```
Comment (Maurits-de-Groot) [inline pytorch-dsv3-mxfp8/README.md]: we ship with CUDA 13 now right? would we either change this to work with CUDA 13 or (given the time limit) just call this CUDA 12.8**+**

File: pytorch-dsv3-mxfp8/README.md [hunk 4]
```diff
@@ -0,0 +1,73 @@
+# DeepSeek-V3 Pre-training with MXFP8 and DeepEP on B200 GPUs
+
+This repository contains the Slurm job scripts and reproducibility instructions for pre-training DeepSeek-V3 (16B and 671B) on a Soperator 256-GPU B200 cluster with TorchTitan.
+
+The recipes include MXFP8 mixed-precision training via TorchAO and DeepEP-accelerated expert communication, which together deliver up to 41% faster training throughput over BF16 without degrading convergence. For the full results and technical deep-dive, see the accompanying PyTorch blog post **[[Link Coming Soon]()]**.
```
Comment (Maurits-de-Groot) [inline pytorch-dsv3-mxfp8/README.md]: nitpick: maybe we can add tokens / sec after 41% faster training throughput to have a unit of measurement included

Now: deliver up to 41% faster training throughput over BF16 without degrading convergence.

Suggestion: deliver up to 41% faster training throughput (tokens / sec) over BF16 without degrading convergence.

File: pytorch-dsv3-mxfp8/README.md [hunk 5]
```diff
@@ -0,0 +1,73 @@
...
+
+## Prerequisites
+
+- Access to a Nebius Soperator cluster:
+  - 32 nodes, 8x NVIDIA B200 GPUs per node, NVLink/NVSwitch and InfiniBand interconnects
+  - Shared filesystem mounted at `/mnt/data`
```
Comment (Maurits-de-Groot) [inline pytorch-dsv3-mxfp8/README.md]: How much space do we need/use of this filesystem?

File: pytorch-dsv3-mxfp8/requirements.txt [hunk 6]
```diff
@@ -0,0 +1,28 @@
+--extra-index-url https://download.pytorch.org/whl/nightly/cu128
+
+# NVIDIA CUDA runtime packages
```
Comment (Maurits-de-Groot) [inline pytorch-dsv3-mxfp8/requirements.txt:3]: see older comment for CUDA 13 question - this could also be done post publication and is not blocking

File: pytorch-dsv3-mxfp8/setup.sh [hunk 7]
```diff
@@ -0,0 +1,23 @@
...
+
+if [ ! -d "torchtitan" ]; then
+    git clone https://github.com/pytorch/torchtitan.git
+fi
+cd torchtitan && git checkout 1a36996b8dffb464340c280d3f0a8139ca4c36b6  # Nightly accessed 2026-02-03
+pip install -e .                
+
+# Download tokenizers
+python3 -c "
+from huggingface_hub import snapshot_download
```
Comment (Maurits-de-Groot) [inline pytorch-dsv3-mxfp8/setup.sh:18]: Where is this dependency coming from? I did not see it in the `requirements.txt`

File: pytorch-dsv3-mxfp8/setup.sh [hunk 8]
```diff
@@ -0,0 +1,23 @@
...
+snapshot_download('deepseek-ai/deepseek-moe-16b-base', local_dir='assets/hf/deepseek-moe-16b-base', allow_patterns=['tokenizer*', 'special_tokens*'])
+snapshot_download('deepseek-ai/DeepSeek-V3.1-Base', local_dir='assets/hf/DeepSeek-V3.1-Base', allow_patterns=['tokenizer*', 'special_tokens*'])
+"
+cd "$ROOT_DIR"
+mkdir -p slurm_logs
```
Comment (Maurits-de-Groot) [inline pytorch-dsv3-mxfp8/setup.sh:23]: You later have `#SBATCH --output=/home/%u/slurm_logs/%x-%j.out` which is a different dir. Is this dir actually used?

File: pytorch-dsv3-mxfp8/setup.sh [hunk 9]
```diff
@@ -0,0 +1,23 @@
+#!/bin/bash
+set -e
```
Comment (Maurits-de-Groot) [inline pytorch-dsv3-mxfp8/setup.sh]: consider setting `set -euo pipefail`

File: pytorch-dsv3-mxfp8/setup.sh [hunk 10]
```diff
@@ -0,0 +1,23 @@
...
+if [ ! -d "torchtitan" ]; then
+    git clone https://github.com/pytorch/torchtitan.git
+fi
+cd torchtitan && git checkout 1a36996b8dffb464340c280d3f0a8139ca4c36b6  # Nightly accessed 2026-02-03
+pip install -e .                
+
+# Download tokenizers
+python3 -c "
+from huggingface_hub import snapshot_download
+snapshot_download('deepseek-ai/deepseek-moe-16b-base', local_dir='assets/hf/deepseek-moe-16b-base', allow_patterns=['tokenizer*', 'special_tokens*'])
```
Comment (Maurits-de-Groot) [inline pytorch-dsv3-mxfp8/setup.sh]: is posible please add `, revision=<the_version_you're_using>` to help with reproducibility

Comment (Maurits-de-Groot) [reply pytorch-dsv3-mxfp8/setup.sh]: for more info see https://huggingface.co/docs/huggingface_hub/guides/download#from-specific-version

File: pytorch-dsv3-mxfp8/train_16b.sh [hunk 11]
```diff
@@ -0,0 +1,84 @@
...
+    --metrics.log_freq=1 \
+    --training.steps="${STEPS}" \
+    --parallelism.data_parallel_replicate_degree="${DP_REPL}" \
+    --parallelism.data_parallel_shard_degree="${DP_SHARD}" \
+    --parallelism.expert_parallel_degree="${EP}" \
+    --parallelism.tensor_parallel_degree="${TP}" \
+    --parallelism.expert_tensor_parallel_degree="${ETP}" \
+    --training.local_batch_size="${BS}" \
+    --training.seq_len="${SEQLEN}" \
+    --lr_scheduler.warmup_steps=2000 \
```
Comment (Maurits-de-Groot) [inline pytorch-dsv3-mxfp8/train_16b.sh:75]: in line 35 you set 1500 steps, why do we warm up for 2000 steps?

File: pytorch-dsv3-mxfp8/train_16b.sh [hunk 12]
```diff
@@ -0,0 +1,84 @@
...
+        --quantize.linear.mx.recipe_name="mxfp8_cublas_rceil"
+        --quantize.linear.mx.filter_fqns="output,router.gate,wq,wkv,wo,feed_forward.w2,shared_experts"
+    )
+fi
+
+# Launch
+srun torchrun \
+    --nnodes "${SLURM_JOB_NUM_NODES}" \
+    --nproc_per_node 8 \
+    --rdzv_id 101 \
```
Comment (Maurits-de-Groot) [inline pytorch-dsv3-mxfp8/train_16b.sh]: I'm not fond of hardcoding rendezvous IDs since this can lead to odd edge cases.  Can we at least have different IDs for the 16b and 671b job and preferably use something like the SLURM_JOB_ID to make this unique

File: pytorch-dsv3-mxfp8/train_16b.sh [hunk 13]
```diff
@@ -0,0 +1,84 @@
...
+    --rdzv_backend c10d \
+    --rdzv_endpoint "${MASTER_ADDR}:29501" \
+    -m torchtitan.train \
+    --job.config_file "${CONFIG_FILE}" "$@" \
```
Comment (Maurits-de-Groot) [inline pytorch-dsv3-mxfp8/train_16b.sh]: Normally parameter expansion (the `"$@"`) is placed at the end of the string to ensure that the overwrites we're doing are taking effect. If we set something like `--training.steps="${STEPS}"` (line 66) after this we will overwrite what is provided in job.config_file

if this is desired behaviour, great - no action needed.

if we want our config to be the source of truth this could create bugs / unexpected behaviour

File: pytorch-dsv3-mxfp8/train_16b.sh [hunk 14]
```diff
@@ -0,0 +1,84 @@
+#!/bin/bash
+
+#SBATCH --job-name=torchtitan_DSV3_16B_AO
+#SBATCH --output=/home/%u/slurm_logs/%x-%j.out
+#SBATCH --error=/home/%u/slurm_logs/%x-%j.err
+#SBATCH --ntasks=32
+#SBATCH --nodes=32
```
Comment (Maurits-de-Groot) [inline pytorch-dsv3-mxfp8/train_16b.sh:8]: do we need 32 nodes for the 16b model?

File: pytorch-dsv3-mxfp8/train_671b.sh [hunk 15]
```diff
@@ -0,0 +1,90 @@
...
+# Environment settings
+MASTER_ADDR=$(scontrol show hostnames "$SLURM_JOB_NODELIST" | sort -r | head -n 1)
+export MASTER_ADDR=$(getent ahostsv4 "$MASTER_ADDR" | head -n 1 | cut -d " " -f 1)
+echo "MASTER_ADDR: $MASTER_ADDR"
+
+export UCX_NET_DEVICES=ens7
+export NCCL_IB_HCA=mlx5
+
+# NVSHMEM / DeepEP settings (required when EP_BACKEND=deepep)
+export NVSHMEM_REMOTE_TRANSPORT=ibrc
```
Comment (Maurits-de-Groot) [inline pytorch-dsv3-mxfp8/train_671b.sh]: since this is only needed when `EP_BACKEND=deepep` you could set it like:

```bash
if [ "$EP_BACKEND" = "deepep" ]; then
  export NVSHMEM_REMOTE_TRANSPORT=ibrc
  export NVSHMEM_IB_ENABLE_IBGDA=1
  export NVSHMEM_IBGDA_NIC_HANDLER=gpu
fi
```

File: pytorch-dsv3-mxfp8/train_671b.sh [hunk 16]
```diff
@@ -0,0 +1,90 @@
+#!/bin/bash
+
+#SBATCH --job-name=torchtitan_DSV3_671B_AO
+#SBATCH --output=/home/%u/slurm_logs/%x-%j.out
```
Comment (Maurits-de-Groot) [inline pytorch-dsv3-mxfp8/train_671b.sh]: consider either setting this to the slurm_logs dir you created earlier or removing that dir

File: pytorch-dsv3-mxfp8/train_671b.sh [hunk 17]
```diff
@@ -0,0 +1,90 @@
+#!/bin/bash
+
```
Comment (Maurits-de-Groot) [inline pytorch-dsv3-mxfp8/train_671b.sh:3]: consider adding `set -euo pipefail`

File: pytorch-dsv3-mxfp8/train_16b.sh [hunk 18]
```diff
@@ -0,0 +1,84 @@
+#!/bin/bash
+
```
Comment (Maurits-de-Groot) [inline pytorch-dsv3-mxfp8/train_16b.sh:3]: consider adding `set -euo pipefail`

File: pytorch-dsv3-mxfp8/train_671b.sh [hunk 19]
```diff
@@ -0,0 +1,90 @@
...
+# Training hyperparameters
+CONFIG_FILE="torchtitan/models/deepseek_v3/train_configs/deepseek_v3_671b.toml"
+TP=2
+PP=2
+EP=32
+ETP=1
+DP_REPL=1
+DP_SHARD=-1
+LBS=64
+STEPS=20
```
Comment (Maurits-de-Groot) [inline pytorch-dsv3-mxfp8/train_671b.sh:45]: here we do 20 steps and 2000 warmup steps (L80)

File: pytorch-dsv3-mxfp8/train_671b.sh [hunk 20]
```diff
@@ -0,0 +1,90 @@
...
+#SBATCH --nodes=32
+#SBATCH --gpus-per-node=8
+#SBATCH --cpus-per-task=96
+
+# Select a recipe and expert-parallel backend via environment variables:
+RECIPE=${RECIPE:-"bf16"}          # bf16 | mxfp8_default
+EP_BACKEND=${EP_BACKEND:-"standard"}  # standard | deepep
+
+SCRIPT_DIR=$(cd "$(dirname "${BASH_SOURCE[0]}")" && pwd)
+DUMP_FOLDER="$SCRIPT_DIR/outputs_671b_${RECIPE}_${EP_BACKEND}"
```
Comment (Maurits-de-Groot) [inline pytorch-dsv3-mxfp8/train_671b.sh]: This will always dump to the same folder right? Could we maybe add the slurm ID to prevent overlap

**Review by Maurits-de-Groot — CHANGES_REQUESTED** [review]

File: deepep/README.md (hunk 1 above)
Comment (HoomanRamezani) [reply deepep/README.md]: done

File: pytorch-dsv3-mxfp8/README.md (hunk 5 above)
Comment (HoomanRamezani) [reply pytorch-dsv3-mxfp8/README.md]: The filesystem could likely get by with a couple terabytes at minimum but it realistically depends on user data, not sure if we should specify an exact value. also checkpointing will be faster based on the size of the filesystem.

Comment (HoomanRamezani) [reply pytorch-dsv3-mxfp8/README.md]: Will add `minimum 2tb` and resolve this

File: pytorch-dsv3-mxfp8/README.md (hunk 3 above)
Comment (HoomanRamezani) [reply pytorch-dsv3-mxfp8/README.md]: The reported results were ran on CUDA 12.8 and we have no way to run this on CUDA13.0 anyway, I'll add `+`, it likely would be faster on 13 (shame we didn't have it)

File: pytorch-dsv3-mxfp8/setup.sh (hunk 9 above)
Comment (HoomanRamezani) [reply pytorch-dsv3-mxfp8/setup.sh]: thanks!

File: deepep/README.md (hunk 2 above)
Comment (HoomanRamezani) [reply deepep/README.md:76]: I think install runs build if its not built yet, its a bit more verbose this way, and tests pass

File: pytorch-dsv3-mxfp8/setup.sh (hunk 7 above)
Comment (HoomanRamezani) [reply pytorch-dsv3-mxfp8/setup.sh:18]: its a torchtitan dependency and gets installed when you setup torchtitan automatically, I don't think we need to pin it

File: pytorch-dsv3-mxfp8/setup.sh (hunk 10 above)
Comment (HoomanRamezani) [reply pytorch-dsv3-mxfp8/setup.sh]: Done

File: pytorch-dsv3-mxfp8/setup.sh (hunk 8 above)
Comment (HoomanRamezani) [reply pytorch-dsv3-mxfp8/setup.sh:23]: I updated it to be a relative paths, all are aligned I believe now, nice catch

File: pytorch-dsv3-mxfp8/train_671b.sh (hunk 20 above)
Comment (HoomanRamezani) [reply pytorch-dsv3-mxfp8/train_671b.sh]: ✅

File: pytorch-dsv3-mxfp8/train_16b.sh (hunk 18 above)
Comment (HoomanRamezani) [reply pytorch-dsv3-mxfp8/train_16b.sh:3]: ✅

File: pytorch-dsv3-mxfp8/train_671b.sh (hunk 19 above)
Comment (HoomanRamezani) [reply pytorch-dsv3-mxfp8/train_671b.sh:45]: Its a throughput benchmark, so small amount of steps after a large warmup. 
Also these configs are just to match the results from the blog post

File: pytorch-dsv3-mxfp8/train_16b.sh (hunk 14 above)
Comment (HoomanRamezani) [reply pytorch-dsv3-mxfp8/train_16b.sh:8]: You don't need 32 to run it, but this repo is to reproduce blog post results, so yes

File: pytorch-dsv3-mxfp8/train_671b.sh (hunk 15 above)
Comment (HoomanRamezani) [reply pytorch-dsv3-mxfp8/train_671b.sh]: yes, nice one

File: pytorch-dsv3-mxfp8/train_16b.sh (hunk 12 above)
Comment (HoomanRamezani) [reply pytorch-dsv3-mxfp8/train_16b.sh]: Changed this to SLURM_JOB_ID

File: pytorch-dsv3-mxfp8/train_16b.sh (hunk 11 above)
Comment (HoomanRamezani) [reply pytorch-dsv3-mxfp8/train_16b.sh:75]: Here we are just documenting exact configs used to get the results in the blog post

File: pytorch-dsv3-mxfp8/train_16b.sh (hunk 13 above)
Comment (HoomanRamezani) [reply pytorch-dsv3-mxfp8/train_16b.sh]: Great catch, moving it to the end

File: pytorch-dsv3-mxfp8/train_671b.sh (hunk 16 above)
Comment (HoomanRamezani) [reply pytorch-dsv3-mxfp8/train_671b.sh]: fixed

**Review by Maurits-de-Groot — APPROVED** [review]
Thank you for addressing the comments @HoomanRamezani
